crates/sui-core/src/authority/execution_time_estimator.rs

```diff
@@ -0,0 +1,777 @@
+                        moving_average: SingleSumSMA::from_zero(Duration::ZERO),
+                        last_shared: None,
+                    });
+            local_observation
+                .moving_average
+                .add_sample(command_duration);
+
+            // Send a new observation through consensus if our current moving average
+            // differs too much from the last one we shared.
+            // TODO: Consider only sharing observations for entrypoints with congestion.
+            let new_average = local_observation.moving_average.get_average();
+            if local_observation.last_shared.map_or(true, |last_shared| {
+                let diff = last_shared.abs_diff(new_average);
+                diff > new_average.mul_f64(OBSERVATION_SHARING_DIFF_THRESHOLD)
+            }) {
+                debug!("sharing new execution time observation for {key:?}: {new_average:?}");
+                to_share.push((key, new_average));
+                local_observation.last_shared = Some(new_average);
+            }
+        }
+
+        // Share new observations.
+        if !to_share.is_empty() {
+            if let Some(epoch_store) = self.epoch_store.upgrade() {
+                let epoch_store = epoch_store.clone();
+                let transaction = ConsensusTransaction::new_execution_time_observation(
+                    ExecutionTimeObservation::new(epoch_store.name, to_share),
+                );
+                // TODO: Add metrics for shared observations.
+                if let Err(e) = self
+                    .consensus_adapter
+                    .submit_to_consensus(&[transaction], &epoch_store)
+                    .await
+                {
+                    if !matches!(e, SuiError::EpochEnded(_)) {
+                        warn!("failed to submit execution time observation: {e:?}");
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Tracks global execution time observations provided by validators from consensus
+// and computes deterministic per-command estimates for use in congestion control.
+pub struct ExecutionTimeEstimator {
+    committee: Arc<Committee>,
+
+    consensus_observations: HashMap<ExecutionTimeObservationKey, ConsensusObservations>,
+}
+
+#[derive(Debug, Clone)]
+pub struct ConsensusObservations {
+    observations: Vec<(u64 /* generation */, Option<Duration>)>, // keyed by authority index
+    stake_weighted_median: Duration,                             // cached value
+}
+
+impl ConsensusObservations {
+    fn update_stake_weighted_median(&mut self, committee: &Committee) {
+        let mut stake_with_observations = 0;
+        let sorted_observations: Vec<_> = self
+            .observations
+            .iter()
+            .enumerate()
+            .filter_map(|(i, (_, duration))| {
+                duration.map(|duration| {
+                    let authority_index: AuthorityIndex = i.try_into().unwrap();
```

Comment:
`i as u32` seems fine here too

Reply:
almost certainly you are right, but it seems better as a default to use the safe version unless there is a reason not to?